gitpod-io/gitpod
# added validation checks for the werft API

## Description

This adds some validation test to the werft.ts API.

- It checks if a phase is already present with the same name.
- A list is added that contains information about the phases and their slices.
- It checks if a slice has already been closed when a log entry is created with the same name.
- It checks if all slices inside a phase has been closed when the phase is closed.

UPDATE (2022-04-06): This is also fixing https://github.com/gitpod-io/gitpod/issues/9100 by ending phases explicitly. I made it optional adding the name of the phase to end it.

## Related Issue(s)

Fixes #8707 #9100

## How to test

You can test the changes by violating a check that has been implemented by this commit.
Here are some examples:
This commit does not have any problems: https://werft.gitpod-dev.com/job/gitpod-build-wth-test.174/raw
This commit does not close all slices: https://werft.gitpod-dev.com/job/gitpod-build-wth-test.175/raw
This commit adds a log entry when the slice has already been closed: https://werft.gitpod-dev.com/job/gitpod-build-wth-test.176/raw
This commit creates a phase that has already been created: https://werft.gitpod-dev.com/job/gitpod-build-wth-test.178/raw

## Release Notes

```release-note
None
```

## Documentation
